src/Time/BoundaryHistory.hpp

```diff
@@ -0,0 +1,310 @@
+/// \tparam FluxVars type of variables passed to the boundary coupling
+template <typename BoundaryVars, typename FluxVars>
+class BoundaryHistory {
+ public:
+  using local_iterator = BoundaryHistoryIterator<true, BoundaryVars, FluxVars>;
+  using remote_iterator =
+      BoundaryHistoryIterator<false, BoundaryVars, FluxVars>;
+
+  // No copying because of the pointers in the cache.  Moving is fine
+  // because we also move the container being pointed into and maps
+  // guarantee that this doesn't invalidate pointers.
+  BoundaryHistory() = default;
+  BoundaryHistory(const BoundaryHistory&) = delete;
+  BoundaryHistory(BoundaryHistory&&) = default;
+  BoundaryHistory& operator=(const BoundaryHistory&) = delete;
+  BoundaryHistory& operator=(BoundaryHistory&&) = default;
+  ~BoundaryHistory() = default;
+
+  /// Modify history data on one of the sides.  See the corresponding
+  /// method in History for details.
+  //@{
+  void local_insert(Time time, BoundaryVars value, FluxVars deriv) noexcept {
+    local_data_.emplace_back(std::move(time), std::move(value),
+                             std::move(deriv));
+  }
+  void remote_insert(Time time, FluxVars deriv) noexcept {
+    remote_data_.emplace_back(std::move(time), std::move(deriv));
+  }
+
+  void local_insert_initial(Time time, BoundaryVars value,
+                            FluxVars deriv) noexcept {
+    local_data_.emplace_front(std::move(time), std::move(value),
+                              std::move(deriv));
+  }
+  void remote_insert_initial(Time time, FluxVars deriv) noexcept {
+    remote_data_.emplace_front(std::move(time), std::move(deriv));
+  }
+
+  void local_mark_unneeded(const local_iterator& first_needed) noexcept;
+  void remote_mark_unneeded(const remote_iterator& first_needed) noexcept;
+
+  local_iterator local_begin() const noexcept { return local_data_.begin(); }
+  local_iterator local_end() const noexcept { return local_data_.end(); }
+
+  remote_iterator remote_begin() const noexcept { return remote_data_.begin(); }
+  remote_iterator remote_end() const noexcept { return remote_data_.end(); }
+
+  size_t local_size() const noexcept { return local_data_.size(); }
+  size_t remote_size() const noexcept { return remote_data_.size(); }
+  //@}
+
+  /// Evaluate the coupling function at the given local and remote
+  /// history entries.  The coupling function will be passed the local
+  /// and remote FluxVars and should return a BoundaryVars.  Values
+  /// are cached internally, so callers should ensure that the
+  /// coupling functions provided on separate calls produce the same
+  /// result.
+  template <typename Coupling>
+  const BoundaryVars& coupling(Coupling&& c, const local_iterator& local,
+                               const remote_iterator& remote) noexcept;
+
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+ private:
+  template <size_t Side>
+  void clean_cache(const Time* const goner) noexcept;
```

Comment:
I'd prefer a better name than `goner` as it is a bit informal/slang, perhaps `to_remove` or `to_be_removed`